File: packages/health_connector_hc_android/android/src/test/kotlin/com/phamtunglam/health_connector_hc_android/unit_tests/mappers/health_measurement_unit_mappers/VelocityMapperTest.kt
package com.phamtunglam.health_connector_hc_android.unit_tests.mappers.health_measurement_unit_mappers

import androidx.health.connect.client.units.Velocity
import com.phamtunglam.health_connector_hc_android.mappers.health_measurement_unit_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_measurement_unit_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.VelocityDto
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

/**
 * Unit tests for Velocity Mapper.
 *
 * Tests verify proper bidirectional mapping between [VelocityDto] and Health Connect
 * [Velocity] objects.
 */
@DisplayName("VelocityMapper")
class VelocityMapperTest {

    @Test
    @DisplayName(
        "GIVEN VelocityDto in meters/second → " +
            "WHEN toHealthConnect called → " +
            "THEN creates Velocity with correct value",
    )
    fun whenVelocityDtoInMetersPerSecond_thenCreatesCorrectVelocity() {
        // Given
        val dto = VelocityDto(metersPerSecond = 10.0)

        // When
        val result = dto.toHealthConnect()

        // Then
        result.inMetersPerSecond shouldBe 10.0
    }

    @Test
    @DisplayName(
        "GIVEN Velocity object → " +
            "WHEN toDto called → " +
            "THEN converts to DTO in meters/second",
    )
    fun whenVelocityToDto_thenUsesMetersPerSecond() {
        // Given
        val velocity = Velocity.metersPerSecond(10.0)

        // When
        val result = velocity.toDto()

        // Then
        result.metersPerSecond shouldBe 10.0
    }
}

File: packages/health_connector_hc_android/test/unit_tests/src/mappers/measurement_unit_mappers/velocity_mapper_test.dart
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hc_android/src/mappers/measurement_unit_mappers/velocity_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart';

import 'package:test/test.dart';

void main() {
  group(
    'VelocityMapper',
    () {
      group(
        'VelocityToDto',
        () {
          test(
            'converts Velocity to VelocityDto in metersPerSecond',
            () {
              const velocity = Velocity.metersPerSecond(10.0);
              final dto = velocity.toDto();

              expect(dto.metersPerSecond, 10.0);
            },
          );

          test(
            'converts Velocity from kilometersPerHour to metersPerSecond',
            () {
              const velocity = Velocity.kilometersPerHour(36.0);
              final dto = velocity.toDto();

              expect(dto.metersPerSecond, 10.0);
            },
          );
        },
      );

      group(
        'VelocityDtoToDomain',
        () {
          test(
            'maps VelocityDto to Velocity',
            () {
              final dto = VelocityDto(metersPerSecond: 10.0);
              final velocity = dto.toDomain();

              expect(velocity.inMetersPerSecond, 10.0);
            },
          );
        },
      );
    },
  );
}
